scipy/spatial/transform/rotation.py

```diff
@@ -1428,6 +1429,36 @@ def inv(self):
             quat = quat[0]
         return self.__class__(quat, normalized=True, copy=False)
 
+    def magnitude(self):
+        """Get the magnitude(s) of the rotation(s).
+
+        Returns
+        -------
+        magnitude : ndarray or float
+            Angle(s) in radians, float if object contains a single rotation
+            and ndarray if object contains multiple rotations.
+
+        Examples
+        --------
+        >>> from scipy.spatial.transform import Rotation as R
+        >>> r = R.from_quat(np.eye(4))
+        >>> r.magnitude()
+        array([3.14159265, 3.14159265, 3.14159265, 0.        ])
+
+        Magnitude of a single rotation:
+
+        >>> r[0].magnitude()
+        3.141592653589793
+        """
+
+        cosines = self._quat.reshape((len(self), 4))[:, 3]
```

Comment:
I believe there is a reason why it was implemented like

```
        quat = self._quat.copy()
        # w > 0 to ensure 0 <= angle <= pi
        quat[quat[:, 3] < 0] *= -1

        angle = 2 * np.arctan2(np.linalg.norm(quat[:, :3], axis=1), quat[:, 3])
```
`arctan2` has uniform numerical properties for all angles. For example, is we use `arccos` for small angles we have `cos x = 1 - x^2/2`, thus for `x ~< sqrt(EPS)` we will get `cos x = 1` and `x = 0`. 

So I suggest that we should use `arctan2` here as well. The line `quat[quat[:, 3] < 0] *= -1` should be replaced by taking `abs` of `arctan2`.

Reply:
Thanks for the review @nmayorov.  I have made the changes you suggested, with the small difference of taking the `abs` of the cosine, rather than the `arctan2`.  I have tested the correctness on a few million random samples.